File: sorting/merge-sort/js/merge-sort.js
const arr = [7, 5, 8, 12, 10, 2, 9, 10];

function mergeSorting(arr, start, end) {
  if (start >= end) return;

  let mid = Math.floor((start + end) / 2);

  mergeSorting(arr, start, mid);
  mergeSorting(arr, mid + 1, end);
  mergeArray(arr, start, mid, end);
}

function mergeArray(arr, start, mid, end) {
  let left = start;
  let right = mid + 1;
  let merged = [];

  while (left <= mid && right <= end) {
    if (arr[left] < arr[right]) {
      merged.push(arr[left]);
      left++;
    } else {
      merged.push(arr[right]);
      right++;
    }
  }

  while (left <= mid) {
    merged.push(arr[left]);
    left++;
  }

  while (right <= end) {
    merged.push(arr[right]);
    right++;
  }

  for (let i = 0; i < merged.length; i++) {
    arr[start + i] = merged[i];
  }
}

mergeSorting(arr, 0, arr.length - 1);
console.log(arr); // Output: [2, 5, 7, 8, 9, 10, 12]

File: sorting/merge-sort/py/merge-sort.py


arr = [7, 5, 8, 12, 10, 2, 9, 10]


def merge_sorting(arr,start,end):
    
    if start >= end : return

    mid = (start+end)//2

    merge_sorting(arr,start,mid)
    merge_sorting(arr,mid+1,end)
    merge_array(arr,start,mid,end)


def merge_array(arr,start,mid,end):
    
    left = start
    right = mid+1
    merged = []

    while (left <= mid and right <= end):
          
          if arr[left] < arr[right]:
            merged.append(arr[left])
            left += 1
          else:
            merged.append(arr[right])
            right += 1

    
    while left <= mid:
      merged.append(arr[left])
      left += 1

    while right <= end:
      merged.append(arr[right])
      right += 1


    for i in range(0,len(merged)):
       arr[start+i] = merged[i]



merge_sorting(arr,0,len(arr)-1)
print(arr)
     


    
